[1단계 - 자동차 경주 구현] 쿤(홍동건) 미션 제출합니다.

Hongdonggeon opened the pull request:
안녕하세요 로운님! 쿤입니다. 

로운님께 코드리뷰를 받을 수 있게 되어 너무 영광입니다 🤘🏻

포키님과 자동차 경주 구현 TDD 페어프로그래밍 미션을 진행하면서 의문이 생겼던 부분에 대해서 몇가지 질문을 드리고 싶습니다!!



### 1. Static 메서드를 사용해야 할 때와 인스턴스 메서드를 사용해야 할 때

메서드를 선언을 할 때 static 메서드로 선언을 해야하는 경우와 인스턴스 메서드로 선언을 해야 하는 경우에 대해 여쭤보고 싶습니다.

저 같은 경우에는 코드 설계를 하면서 Car 같이 여러 객체가 생성되는 경우와 그리고 멤버 필드가 있는 클래스 안에서는 보통 인스턴스 메서드로 선언을 하고 있습니다. 

반대로 static 메서드는 형변환을 해주거나, 객체를 생성해주는 객체가 직접 하는 일이 아닌 경우에 static 메서드를 선언을 하고 있습니다. 또한 멤버 필드가 없는 클래스에서는 메서드들을 static 메서드로 해도 되지 않을까라는 의문이 항상 생기는 것 같습니다.



### 2. 테스트 코드의 범위

TDD를 처음 경험을 해보는데요! 

직접 기능을 가지고 있는 클래스 이외에 모든 코드를 조립하고 흐름을 제어해주는 메서드는 먼저 테스트 코드를 작성하는 방법이 떠오르지 않았습니다. 

제가 작성한 코드의 예시로는 start(), showResult(), showWinner() 같은 메서드들인데요! 반환형이 void 형이기도 하고 테스트 코드 시작을 어떻게 작성을 해야할지 궁금합니다. 그리고 이런 메서드들도 테스트 코드의 필요성이 있는지도 궁금합니다.



### 3. toString()

ui로직과 도메인 로직의 독립적인 코드에 대해 질문 드리고 싶습니다.

자동차 경주 결과 발표 로직을 작성할 때 'kun : ---' 처럼 출력하기 위한 방법으로 2가지 방법이 떠올랐는데요! 

첫번째는 toString() 을 오버라이딩 하여 코드를 설계하는 것이었는데요! 이 방법을 사용해서 구현을 하였습니다.

하지만 걱정이 되는 부분이 있다면 toString()은 객체를 표현을 해주는 메서드라고 생각을 하기 때문에, 다음에 더 중요한 정보나 더 포괄적인 정보를 표현하도록 요구사항이 생긴다면 전부 리팩토링을 해야하니까 toString()을 오버라이딩해서 사용하는게 권장되는지 궁금합니다.

두번째는 CarDto라는 클래스를 생성을 하여, Game 클래스에서 Car를 CarDto로 형변환 후 OutputView에 넘겨주는 방법을 생각을 해봤습니다. 이 방법 같은 경우에는 형변환을 하는 코드가 더 많아진다고 생각해 더 간결한 첫번째 방법을 사용을 해보았습니다. 어떤 방법이 더 권장되는지 궁금하고, 또 다른 방법이 궁금합니다.

### 4. 커밋 습관

이번에 페어프로그래밍하면서 포키님께 너무 좋은 커밋 습관을 배운 것 같다고 생각을 합니다!! 

저는 지금 까지 커밋을 그렇게 세세하게 남기지 않았었는데, body footer까지 커밋을 기록하는 연습을 해보았습니다.

새로운 개념을 적용하는 경우에는 커밋 로그를 기록을 하면서 한번 더 정리하는 느낌도 가질 수 있었는데요!!

좋은 커밋 습관의 팁이 있거나, 실제 실무에서도 이렇게 커밋 기록을 자세하게 남기는지 궁금해서 질문드립니다!!

<hr/>
아래 링크는 포키님과 저와 공유한 노션 링크입니다!!

https://brazen-serpent-b4f.notion.site/Lv1-Racing-car-47749ac1c64945f797a0506f4ee0eff9

YJGwon committed:
test: String API에 대한 학습 Test class 추가

junitstudy/StringTest.java 추가
   - split_구분자_있을때(), split_구분자_없을때(): 구분자 단위 문자열 분리 test, containsExactly()는 배열의 모든 값이 일치해야 통과
   - subString_length_소괄호_제거(): length()를 이용해서 input값의 길이가 달라져도 항상 맨 앞, 맨 끝 문자를 제거
   - subString_index_소괄호_제거(): indexOf()를 이용해서 input값의 형식이 달라져도 항상 여는 괄호와 닫는 괄호를 제거
   - charAt_특정위치_문자_반환(): charAt()의 return 값을 String.valueOf()로 형변환
   - indexOutOfBoundsException_발생(): 발생한 Exception의 type이 IndexOutOfBoundsException인지 검사
   - exception_message_검사(): withMessage()를 통해 error message 일치 검사

YJGwon committed:
test: Set API에 대한 학습 Test class 추가

juniitstudy/SetTest.java 추가
   - set_size_확인(): set은 중복 제거를 해주기 때문에 중복 된 값 한번만 저장
   - set값_전체_존재_확인(int value): @ParameterizedTest와 @ValueSource를 이용해 여러개의 값을 파라미터에 대입해서 실행
   - set값_존재_여부_구별(int value, boolean result): @CsvSource를 이용해 두 파라미터에 여러쌍의 값을 대입, 자료형에 맞게 String값이 파싱된다

YJGwon committed:
feat: 문자열 계산기 최초 commit

YJGwon committed:
docs: stringcalculator\README.md 추가

기능 목록 최초 작성

YJGwon committed:
feat: 입력값이 null값 또는 빈 문자열인지 판별하여 0을 반환

StringCalculator.java
   - splitAndSum() 추가
   - castStringToInts(String input): String 판별해서 int Array로 cast
   - calculate(int[] numbers): cast된 배열의 총합을 계산

YJGwon committed:
docs: 완료된 기능 체크

YJGwon committed:
feat: 커스텀 구분자가 존재하는 경우

StringCaculator.java
   - separate(String input, String separator): 구분자로 문자열을 쪼개서 int array로 반환
   - getCustomSeparatorMatcher(String input): 커스텀 구분자 패턴 matcher 반환

YJGwon committed:
feat: 커스텀 구분자가 없는 경우

StringCalculator.java
   - castStringToInts(): : 과 , 를 구분자로 지정해서 separate() 호출

YJGwon committed:
feat: [예외] 변환한 숫자 배열 중 음수가 있는 경우

StringCalculator.java
   - validatePositive(): 숫자가 음수인 경우 RuntimeException 발생
   - separate(): Stream 중간 연산 메서드 peek을 사용해서 Stream에 영향을 주지 않고 모든 요소에 대해 validatePositive 실행
StringCalculatorTest.java
   - splitAndSum_not_number(): 숫자가 아닌 값일 경우에 Exception에 대한 테스트 미완성

YJGwon committed:
docs: README.md 기능 목록 최초 작성

YJGwon committed:
feat: 주어진 이름으로 Car 객체 생성

Car.java
   - private 멤버 name 추가
CarTest.java
   - 차_생성()

YJGwon committed:
feat: 입력값을 List<Car>로 만들어주는 CarFactory Class 추가

CarFactory.java : 정적 팩토리 패턴 응용

name에 대한 길이 검증 필요

YJGwon committed:
feat: [예외] 이름이 5글자를 초과할 때

Car.java
   - validateName(): 이름 5글자 이하인지 검사
CarTest.java
   - 이름_5글자_이상: assertThatThrownBy의 hasMessageStartingWith을 이용해 에러 메세지 비교

에러 메세지 출력과 재시작 구현 필요

YJGwon committed:
feat: [예외] 이름이 공백일 경우

Car.java
   - validateEmptyName(): 이름이 공백인지 검사
CarTest.java
   - 공백_이름()

에러 메세지 출력과 재시작 구현 필요

YJGwon committed:
feat: [예외] 차 이름이 중복인 경우

Car.java
   - equals(), hashCode(): Car 객체의 이름이 같은 경우는 같은 객체로 취급
CarFactory.java
   - of(): 리턴 타입을 List에서 Set으로 변경해서 collect하면서 바로 중복 제거
   - validateDuplicatedCarName(): 문자열 배열과 Set의 크기가 다를 때 중복 이름 에러 발생
CarTest.java
   - 중복된_이름(): 중복된 이름 에러 발생 테스트

YJGwon committed:
feat: 값이 4 이상일 경우 전진

Referee.java
   - judgeCarMove(Car car, int randomValue): 전달받은 값이 4이상일 경우 전진시킨다
Car.java
   - position: 차의 현재 위치를 나타내는 멤버 변수
   - move(): position에 1을 더한다
RefereeTest.java
   - 차_전진_판단: 4이상의 값을 줬을 때 차가 전진하는지 테스트

YJGwon committed:
feat: 난수 생성하기

Game.java
   - makeRandomValue(): 0부터 9사이의 랜덤 값 생성
GameTest.java
   - 랜덤값_부여_테스트(): 랜덤값이 범위에 맞게 생성하는지 검사

난수 생성은 굳이 테스트 하지 않아도 되지만, 범위에 대해 테스트 하는 좋은 방법을 고민해보기 위해 테스트 코드 작성

YJGwon committed:
feat: 우승자 판정

Referee.java
   - getMaxPosition(Set<Car> cars): stream의 max()를 이용해서 최대 position값 도출
   - judgeWinner(Set<Car> cars): getMaxPosition()에서 도출한 값과 position이 같은 Car 객체 List 반환
RefereeTest.java
   - 우승자_판정(): position값에 차이를 뒀을 때 우승자 판정 테스트
   - 우승자_여러명(): 우승자가 여러명일 때 테스트

YJGwon committed:
feat: InputView 추가

YJGwon committed:
feat: Car.toString 추가

CarTest.java
   - toString_kun(): 이름이 포함되는지
   - toString_forky(): 다른 이름을 넣어도 통과하는지(멤버변수 name 사용)
   - toString_position_1(): position charactor(-)가 포함되는지
   - toString_position_3(): position이 바뀌어도 통과하는지 (멤버변수 position 만큼 position charactor를 StringBuilder로 연결)

TDD 피드백 강의 이후 최소의 구현으로 test를 통과시키는 TDD Cycle을 적용해서 toString()을 완성

YJGwon committed:
feat: 이름, 횟수 입력받아서 게임 1회 진행

Game.java
   - start(): 실제 게임을 시작하고, 게임의 흐름을 제어하는 역할
   - play(): Car객체에 난수를 부여하고 전진하는 게임 진행 1회 수행
   - showResult(): 게임의 1회 진행 결과 출력

YJGwon committed:
feat: 우승자 출력

Game.java
   - showWinner(): Stream의Collectors.joining()을 사용해서 collect가 될 때원하는 형태의 문자열로 결합되게 함

YJGwon committed:
feat: 시도횟수만큼 반복

YJGwon committed:
feat: [예외] 횟수가 음수일 때

InputView.java 오류 fix
   - inputCarNames(): 이름에 공백이 있을 때 inputGameCount가 실행되며 발생하는 버그를 next() -> nextLine()으로 변경하여 해결

Hongdonggeon committed:
docs: README 파일 수정

lowoon commented on src/main/java/racingcar/Game.java:
```diff
@@ -0,0 +1,54 @@
+	}
+
+	private void validateGameCount(int gameCount) {
+		if(gameCount < 0) {
```
0은 상수로 만드는게 낫지 않을까요??

Hongdonggeon replied:
에러 발생하는 기준값을 상수화도 해주어야했네요!! 반영했습니다!

lowoon commented on src/main/java/racingcar/CarFactory.java:
```diff
@@ -0,0 +1,24 @@
+		String[] arrNames = names.split(",");
+		Set<Car> cars = Arrays.stream(arrNames)
+			.map(name -> new Car(name.trim()))
+			.collect(Collectors.toSet());
```
```suggestion
		Set<Car> cars = Arrays.stream(arrNames)
		        .map(String::trim())
			.map(Car::new)
			.collect(Collectors.toSet());
```
가능하면 한번에 여러가지를 하지 말고 step by step으로 하나의 행위 위주로 작성해주는게 깔끔하답니다~

Hongdonggeon replied:
좋은 팁 감사합니다!! 리뷰 반영했습니다!!

lowoon commented on src/main/java/racingcar/Game.java:
```diff
@@ -0,0 +1,54 @@
+	}
+
+	public int makeRandomValue() {
+		Random random = new Random();
```
Random을 항상 새로 생성하여 사용해야 할까요?

Hongdonggeon replied:
오 거기까지는 생각을 못한 것 같아요,,,!
그러면 Random 객체를 Game 클래스에서 인스턴스 변수로 빼서 Game 객체가 한번 만들어질때 Random 객체가 한번 만들어지게 하면 될까요?

lowoon commented on src/main/java/racingcar/Car.java:
```diff
@@ -0,0 +1,76 @@
+		this.name = name;
+	}
+
+	public String getName() {
```
다른 중요 로직 메서드를 위에 두고 getter를 아래에 두는게 낫지 않을까요??

```
public class ClassName {
	// private static final
	// 상수
	// 클래스 변수
	// 인스턴스 변수
	// 생성자
	// 팩토리 메서드
	// 일반 메서드
	// getter, setter
	// equals, hashCode, toString
}
```

Hongdonggeon replied:
리뷰 반영했습니다!

lowoon commented on src/main/java/racingcar/Referee.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class Referee {
```
cars에 대한 관리인지, 규칙에 대한 관리인지, 차량을 움직이는 역할인지 Referee의 역할이 애매한 거 같아요.

referee의 역할에 대해 고민해보고 다른 객체를 생성하거나 좀 더 명확한 역할로 정해보는게 나을 거 같아요.

Hongdonggeon replied:
제가 생각한 Referee의 역할으로는 
- Car 객체가 갈 수 있는지 판정해준다.
- 게임에 참여한 Car 들의 우승자를 골라준다.

전체적인 게임에 대해 판정을 해주는 역할이라고 생각했어요!
우승자를 고르기 위한 작업을 위한 최대값을 찾는 과정도 Referee에 있는게 가장 알맞다고 생각해서 Referee에 추가했었습니다.

리뷰를 보고 다시 생각해보니 Referee가 차를 움직이게 하는 것은 진짜 역할이 애매한 것 같네요,,
그리고 Set<Car>에서 최대값을 골라주는 역할을 하는 Cars 클래스가 있으면 좋을 것 같다는 생각을 해볼 수 있었습니다.
Referee는 게임을 위한 판정이 필요한 부분에 판정만 해주는 역할만 하게 리팩토링 해보겠습니다!

Hongdonggeon replied:
Referee
- Car가 갈수 있는지 판정만
- Set<Car>에서 파라미터로 받은 최대값과 같은 Car들 필터링 (우승자 판정)

Cars
- Cars에 대한 관리

로 역할을 나누어 리팩토링 해보았습니다!!

lowoon replied:
역할에 대해 고민해보고 반영하신 점 너무 좋네요 👍 
referee에 단 코멘트도 고민해보시면 좋을거 같아요

lowoon commented on src/test/java/racingcar/RefereeTest.java:
```diff
@@ -0,0 +1,34 @@
+		Car car = new Car("kun");
+		Referee.judgeCarMove(car, 5);
+		assertThat(car.getPosition()).isEqualTo(1);
+	}
```
테스트를 작성할 때 기준이 되는 값 전,후를 다 테스트 해주는게 좋습니다.
예를 들어 기준이 4라면 4,5 혹은 3,4,5가 되겠죠??

Hongdonggeon replied:
오 좋은팁 감사합니다!! 리뷰 반영했습니다!
Car 이름 길이 테스트도 기준 값인 5와 전, 후로 테스트 코드도 추가했습니다 :)

lowoon reviewed (changes requested):
반가워요 쿤! 리뷰어 로운입니다! 😄
작성하신 코드 잘 확인했습니다. 다양한 방법을 고민해보고 시도해보신 것 같아 좋았습니다 👍
소소한 피드백 몇 가지 남겨봤으니 확인 부탁드려요.

질문하셨던 부분 답변을 드리면
1. static
말씀해주신 것과 같이 static은 상수와 util class에서 많이 쓰이고 멤버변수가 없는 클래스에서도 쓰이는 것 같습니다. 이때 멤버 변수가 없는 클래스의 경우는 멤버변수가 왜 없는지, 이 클래스의 역할과 기능은 뭔지 고민해보고 사용하는게 좋을거 같아요.
참고하시면 좋을거 같아요. https://tecoble.techcourse.co.kr/post/2020-07-16-static-method/

2. 테스트 코드
void의 경우에는 반환값을 넘기는 형태로 수정할 수 없을까 고민을 해보는 것 같고요. 말씀하신 start()의 경우에는 view와 게임의 로직을 분리해서 테스트 해볼 수 있지 않을까요?? 테스트는 내가 작성한 코드가 내가 의도한대로 동작했는지 확인하는 작업입니다. 그렇다면 어디가 제대로 동작했는지, 확인하고 싶으신 부분을 생각해보시면 좋을거 같아요.

3. toString()
toString은 말씀하신 것처럼 객체를 표현해주는 메서드입니다. 이 메서드를 outputView가 아닌 다른 곳에서도 사용을 한다고 하면 지금 상황에서는 사용을 하지 못하겠죠??
DTO객체를 사용하는 것이 권장되는데 DTO객체를 사용하는 이유에 대해서 고민해보시면 좋을거 같아요.

4. 커밋
말씀하신 커밋은 팀마다 굉장히 다른 것 같습니다. 어떤 팀은 굉장히 세세하게 적기도 하고, 어떤 팀은 큰 틀의 제목만 적기도 하는 것 같아요. 기본적으로 커밋이란 어떠한 작업을 했는지, 나 뿐만이 아닌 다른 개발자들도 확인할 수 있는 수단입니다. 함께 일하는 개발자들의 합의에 의해 정해졌다면 어떠한 형태도 괜찮다고 생각합니다.
다만, 추후에 커밋을 통해 확인하는 상황을 고려하고, 말씀하신 것처럼 한번 더 정리하는 느낌을 가지셨다면 자세하게 남기는게 좋지 않을까요??

답변이 도움이 되었으면 좋겠네요 😸 
궁금한 점이 생기면 언제든지 DM 혹은 댓글남주세요 ^-^

lowoon commented on src/main/java/racingcar/Car.java:
```diff
@@ -0,0 +1,76 @@
+	public Car(String name) {
+		validateEmptyName(name);
+		validateName(name);
+		this.name = name;
```
position도 같이 초기화해주는게 좀 더 명확할거 같아요.

Hongdonggeon replied:
인스턴스 변수는 기본값이 0으로 설정이 되어서 초기화를 안해줘도 되지 않을까 생각했어요! 
자동차 경주의 시작점이 바뀔 수도 있고, 0에서 시작한다고 명시해주는 것이 더 명확할 것 같네요!!

lowoon commented on src/test/java/racingcar/CarTest.java:
```diff
@@ -0,0 +1,74 @@
+	public void 이름_5글자_이상() {
+		assertThatThrownBy(() -> new Car("abcdef"))
+			.isInstanceOf(IllegalArgumentException.class)
+			.hasMessageStartingWith(ERROR_MESSAGE);
```
공통적인 에러 메세지가 있는지만 테스트하기보다 실제 상황에 맞는 메세지가 나오는지를 확인하는게 진짜 테스트하고싶은 것 아닐까요??

Hongdonggeon replied:
hasMessage()라는 정확히 에러문과 일치하는 비교해주는 메서드가 있었네요! 
덕분에 hasMessage(), hasMessageContaining(), hasMessageStartsWith()의 차이점에 대해 공부할 수 있었습니다! 
감사합니다😀

Hongdonggeon committed:
refactor: 에러 발생 기준값 상수화

- Game.java
   입력하는 값이 0 이하일 때 예외를 발생하는데, 이 기준 값인 0을 상수화

Hongdonggeon committed:
refactor: 스트림 한 단계 중간 연산 추가

CarFactory.java
   -of(): 한번에 여러가지 작업을 하는 것보다 하나의 행위 위주로 한단계씩 작성하는 것이 좋다.

Hongdonggeon committed:
refactor: 클래스 내 메서드 위치 변화

Hongdonggeon committed:
refactor: position 초기값 0 명시

Car.java
   자동차 경주의 시작점이 바뀔 수 있고, 0에서 시작한다고 명시해주는 것이 더 명확함

Hongdonggeon committed:
refactor: Random 객체 멤버 변수에서 인스턴스 변수로 전달

Game.java
   랜덤값을 생성할 때 계속해서 객체를 만들어주는 것보다 Game 클래스가 만들어질 때 Random 객체 한번 생성

Hongdonggeon committed:
refactor : play() 메서드 반환형 void 로 변환

Game.java
   - play(): showResult(play(cars));로 바꿔줄 수 도 있었지만 지금 처럼 작성한 것이 가독성도 더 좋고 자동차 경주 게임 흐름을 더 잘 표현한다고 생각했다.

Hongdonggeon committed:
refactor: 예외 발생 테스트 코드 에러문 비교하는 메서드 변경

CarTest.java
   - hasMessage(): 발생한 에러의 에러문과 정확히 문자열이 일치하는지 비교
   - hasMessageContaining(): 발생하는 에러문에 파라미터로 들어온 문자열이 포함되는지
   - hasMessageStartsWith() : 발생하는 에러문과 파라미터로 들어온 문자열 처음부터 비교, 정확히 문자열이 일치하지 않아도 되고 초반부터 똑같기만 하면 됨

Hongdonggeon committed:
refactor: judgeCarMove 역할 분리

- Referee.java
  - canCarMove():  이전에는 Referee 객체가 랜덤값을 부여받아 Car객체를 움직여 주었다면, Referee가 랜덤값을 부여받아 Car가 움직일 수 있는지 없는지 판정만 내려준다.
Car.java
   - move(boolean): 이전에는 단지 움직이는 메서드였다면 Referee 객체에게 움직일수있다는 판정을 받으면 움직이기

이외의 변화된 클래스들은 메서드의 구조가 바뀌면서 자연스럽게 바뀌게 되었다.

Hongdonggeon committed:
Test: 랜덤값에 따른 전진여부 판단 테스트 추가

RefereeTest.랜덤값이_3일때_차_전진_판단
RefereeTest.랜덤값이_4일때_차_전진_판단
RefereeTest.랜덤값이_5일때_차_전진_판단
   - 테스트를 작성할 때 기준이 되는 값 전, 후를 다 테스트 해주는것이 좋다.

Hongdonggeon committed:
Test: Car객체 생성 테스트

CarTest.이름_4글자_일때_차_생성
CarTest.이름_5글자_일때_차_생성
CarTest.이름_6글자_일때_차_생성
   - Car 객체의 이름의 최대 길이 기준값인 5와 전 후 테스트

Hongdonggeon committed:
feat: 여러 Car객체를 관리하는 Cars 클래스 추가

Cars.java
   Car객체들을 관리하는 일급 컬렉션, 생성자에 List<Car>를 파라미터로 받아 객체 생성

CarsTest.java
   여러개의 Car객체들이 빠짐없이 Cars 객체의 Set<Car>에 추가되는지 테스트

Hongdonggeon committed:
refactor: Car 객체 생성해주는 팩토리 메서드 역할 분리

CarFactory.java
   of(): 이전에는 문자열을 쪼개서 중복 검사를 해주고 Set으로 반환했다면, 증복검사는 Cars 객체에 위임하고 단지 문자열을 쪼개서 Car객체만 생성해주는 기능만 하도록함

Hongdonggeon committed:
feat: Cars 객체를 만들 때, 중복 Car 객체 검사해주기

Cars.java
   validationDuplicatedCarName(): 스트림의 distinct()를 사용해서 중복 검사해주기
   getCars(): unmodifiableSet()을 이용해서 불변 객체 반환해주기
CarsTest.java
    중복된_이름(): CarTest에서 CarsTest로 이동
CarTest.java
   차_1대_생성(): Car 객체를 만들어주는 CarFactory.of()에 대한 테스트 추가

이외의 변화들은 여러 Car 객체를 관리하고 반환해주는 Cars 클래스를 추가함에 따라 생긴 자연스러운 변화들임

Hongdonggeon committed:
refactor: Car 객체 집합에서 최대값 position 찾는 메서드 위치 이동

Cars.java
   getMaxPosition(): 이전에는 Referee 클래스에 존재하였는데 Set<Car>를 관리해주는 Cars 객체에 역할 위임

Referee.java
   judgeWinner(Set<Car>, int): Referee는 최대값 position을 파라미터로 받아 파라미터로 받는 Set<Car>에서 최대 position 값 과 position이 똑같은 Car 들만 필터링해줌

Game.java
   Cars의 getCars()를 계속해서 호출한 것보다, 한번만 호출하여 사용하는게 더 낫다고 생각

이외의 변화는 메서드 구조의 변화로 자연스럽게 바뀜

Hongdonggeon committed:
feat: Cars의 Set<Car>에서 이름이 같은 Car 반환

Cars.java
   getCarByName(String): Cars의 Set<Car>에서 파라미터로 들어오는 문자열과 이름이 같은 Car 객체 반환

CarsTest.java
   getCarByName()을 이용해 이름이 같은 Car가 잘 반환되는지 테스트

Hongdonggeon committed:
Test: Cars의 getCarByName()과 getMaxPosition() 테스트 코드 추가

CarsTest.Car_집합에서_찾는_이름이_존재_X(): 찾는 이름이 존재하지 않은 경우

lowoon commented on src/main/java/racingcar/Car.java:
```diff
@@ -0,0 +1,79 @@
+        return this.position == position;
+    }
+
+    public void move(boolean canCarMove) {
```
리뷰를 잘 반영해 주셔서 이것을 2단계에 도전해 보죠 😎 
https://tecoble.techcourse.co.kr/post/2020-05-07-appropriate_method_for_test_by_parameter/
https://tecoble.techcourse.co.kr/post/2020-05-17-appropriate_method_for_test_by_interface/

Hongdonggeon replied:
오! 이런 방법이 있네요~~ 너무 많은 것 배워갑니다! 숫자를 반환해주는 인터페이스를 구현하여, 랜덤값을 반환해주는 구현체는 프로덕션 코드에서, 자신이 의도한 숫자를 반환해주는 구현체는 테스트 코드에서 사용한다고 이해하였습니다. 그러면 랜덤값을 잘 만들어주는지의 테스트 코드는 따로 작성할 필요가 없는 건가요??

lowoon replied:
**개인적인 생각인데요.** 있으면 좋은데 꼭 필요하다고 생각하지는 않습니다.
random에서 기본적으로 제공하는 메서드이기 때문에 랜덤을 생성해주는 것에는 이슈가 없을 것이고, 범위만 잘 설정하면 되기 때문에 이 부분은 상수로 표현하여 쓸 수 있다고 생각합니다.

Hongdonggeon replied:
계속 생각해보니 이해가 가는 것 같아요!! Random 클래스 같은 경우는 자바 api이기 때문에 이미 검증?되었다고 생각이 들고, 또한 테스트 하고 싶은 내용이 4를 기준으로 전진하는지 못하는지 이니까 따로 안해도 될 것 같다는 생각이 들었습니다!!

lowoon commented on src/main/java/racingcar/CarFactory.java:
```diff
@@ -0,0 +1,18 @@
+import java.util.stream.Collectors;
+
+public class CarFactory {
+    private static final String DUPLICATED_NAME_ERROR = "[ERROR] 차 이름은 중복될 수 없습니다.";
```
에러 메세지가 사용되지 않는것 같네요?

Hongdonggeon replied:
Cars 클래스를 만들면서 Cars 클래스에서 이 상수를 사용한다고 옮겼는데, 삭제를 안해버렸네요,, 더 유심히 봐야되겠습니다!

lowoon commented on src/main/java/racingcar/Referee.java:
```diff
@@ -0,0 +1,19 @@
+	public static List<Car> judgeWinner(Set<Car> cars, int maxPosition) {
+		return cars.stream()
+			.filter(car -> car.isPosition(maxPosition))
+			.collect(Collectors.toList());
```
cars 객체를 만들었다면 cars객체에게 포지션과 같은 car의 리스트를 달라고 해볼 수 있지 않을까요??
참고하시면 좋을거 같아요. https://tecoble.techcourse.co.kr/post/2020-05-08-First-Class-Collection/

Hongdonggeon replied:
참고 자료보고 너무 많은 도움이 됐습니다!! 만약 꼴등도 뽑는 기능이 추가 된다면, judgeWinner() 와 코드가 중복되는 메서드 하나를 더 구현했어야 했네요!! 일급 컬렉션에게 행위를 위임하고 요청하는 방식으로 바꾸었습니다!! 너무 좋은 것 배워갑니다~!!☺️

lowoon commented on src/main/java/racingcar/Car.java:
```diff
@@ -0,0 +1,79 @@
+        for (int i = 0; i < position; i++) {
+            stringBuilder.append(POSITION_CHARACTOR);
+        }
+        return stringBuilder.toString();
```
position을 기초로 해서 어떻게 보여줄지는 view의 역할이라고 생각하는데 어떻게 생각하시나요??

Hongdonggeon replied:
그렇네요!! Car에서 너무 많은 일을 하고 있네요! View와 도메인을 너무 독립적으로 만들려는 욕심때문에 View를 바보로 만들어버렸네요,,, 리뷰 반영했습니다~:)

lowoon commented on src/main/java/racingcar/Cars.java:
```diff
@@ -0,0 +1,46 @@
+                .filter(car -> car.getName().equals(carName))
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException(NOT_FOUND_CAR_NAME_ERROR));
+    }
```
해당 메서드는 테스트에서만 사용되네요??

Hongdonggeon replied:
네 맞아요 테스트 코드에서만 사용하고 있어요
메서드를 만든 이유로는, 실제 프로덕션 코드에서 Car객체들을 만드는 메서드는 CarFactory의 of, car들을 관리해주는 클래스는 Cars 인데,  아래 코드처럼 테스트 돌리기 보다는 실제 메서드들을 사용하고 싶었어요

```java
    @Test
    public void Car_집합에서_최대_position_값_찾기() {
      //given
      	List<Car> carList = new ArrayList<>();
        Car car1 = new Car("kun");
      	Car car2 = new Car("forky");
      	carList.add(car1);
      	carList.add(car2);
      	Cars cars = new Cars(carList);
      
      //when
        for (int i = 0; i < 5; i++) {
            car1.move(true);
        }
      
      //then
        assertThat(cars.getMaxPosition()).isEqualTo(5);
    }
```
<br/>
그리고 지금은 사용할 일이 없지만,  Cars 객체에서 Car 하나를 반환해서 어떤 작업을 처리해줘야하는 순간이 올때, 

```java
  Set<Car> carSet = cars.getCars();
  String carName = "kun";
  Car needCar;

  for(Car car : carSet) {
       if(car.getName().equals(carName)){
    	     needCar = car;
       }
}

```
이렇게 하는 것보다 Cars에서 문자열을 파라미터로 받아, 문자열과 이름이 일치한 Car객체 하나를 반환해주는 메서드를 만들어주는게 좋지 않을까 생각했습니다!

Hongdonggeon replied:
이 메서드를 사용하지 않고 테스트 코드를 작성하는 방법을 고민하다가 방법이 떠올라서 바꿨습니다!!
```java
    @Test
    public void Car_집합에서_최대_position_값_찾기() {
        Cars cars = new Cars(CarFactory.of("kun,forky"));
        String name = "kun";

        for (Car car : cars.getCars()) {
            if (name.equals(car.getName())) {
                car.move(true);

            }
        }

        assertThat(cars.getMaxPosition()).isEqualTo(1);
    }
```

Hongdonggeon replied:
다시 생각해보니, 괜히 저 메소드를 만들어놨다가 예외상황을 하나 더 늘린 것일 수도 있다는 생각이 드는 것 같습니다! 
지금 당장 프로덕션 코드에서 사용하지 않고, 테스트 코드에서만 사용되는 메서드를 만들어두는 것은 안좋은 선택인가요,,?

lowoon replied:
**이것도 개인적인 생각인데요!**
테스트만을 위한 메서드, 코드를 프로덕션 코드에 만드는 것은 좋지 않다고 생각합니다. 말씀하신 것과 같이 프로덕션과 연관이 없는 곳에서 에러가 발생을 할 수도 있고, 프로덕션 코드의 return 값을 바꾸거나 적용해주신 것처럼 다른 방법을 통해 테스트 할 수 있을 테니까요.

Hongdonggeon replied:
답변 너무 감사드립니다😀  의문점에 대해 하나씩 풀려가니 너무 재밌네요!!

lowoon commented on src/main/java/racingcar/Cars.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.List;
+import java.util.Set;
+
+public class Cars {
```
일급 객체 👍

lowoon commented on src/main/java/racingcar/Car.java:
```diff
@@ -0,0 +1,79 @@
+    @Override
+    public String toString() {
+        return name + TO_STRING_CONJUNCTION + buildStringPosition();
+    }
```
> Returns a string representation of the object

toString은 객체를 나타내는 String입니다. override해서 만들어 주신 메서드에서 return 해주는 값이 이 객체를 나타내나요??

간단한 예로 이 값을 자동차 게임이 아닌 다른 곳에서 사용할 수 있나요??

Hongdonggeon replied:
맞아요 저도 그 부분이 우려되었던 것 같아요!! toString()은 객체를 표현을 해주는 메서드이니까 다음에 더 중요한 정보나 더 포괄적인 정보를 표현하도록 요구사항이 생긴다면 전부 리팩토링을 해야될 것 같아요!! 그리고 콘솔에 보이는 UI로직은 View가 할 일인데 Car에서 너무 많이 해버린 것 같기도 하네요!!

lowoon commented on src/main/java/racingcar/Game.java:
```diff
@@ -0,0 +1,57 @@
+	}
+
+	public void showWinner(List<Car> winnerCars) {
+		String winnerNames = winnerCars.stream()
+			.map(Car::getName).collect(Collectors.joining(WINNER_NAME_DELIMITER));
```
```suggestion
		String winnerNames = winnerCars.stream()
			.map(Car::getName)
			.collect(Collectors.joining(WINNER_NAME_DELIMITER));
```
메서드 체이닝은 한줄에 한개씩 표현해주는 것이 명확하답니다~

Hongdonggeon replied:
넵! 한줄에 한개씩 표현해주는 것, 한단계씩 처리해주는 것! 잊지 않겠습니다!

lowoon commented on src/test/java/racingcar/RefereeTest.java:
```diff
@@ -0,0 +1,44 @@
+		car1.move(true);
+		cars.add(car1);
+		cars.add(car2);
+		assertThat(Referee.judgeWinner(cars, 1)).containsExactly(car1);
```
given when then 형식으로 바꿔볼까요??

Hongdonggeon replied:
넵 바꿨습니니다!! 테스트 코드 작성할 때 주석으로 given when then 명시해주는게 더 좋은가요?

lowoon replied:
given when then은 기본 템플릿으로 개발자들이 알고 있는 내용이라 딱히 명시해줄 필요는 없다고 생각해요.
다만 저 틀에 맞게 작성하기 위해 ide 자체에 템플릿으로 만드는 경우가 있는데 이 경우에 굳이 지울 필요는 없겠죠??

Hongdonggeon replied:
given when then, given when then 계속 연습하겠슴돠~!!

lowoon reviewed (approved):
고생하셨어요. 쿤! 이전 리뷰 내용을 잘 반영해주셨습니다 💯 
몇가지 코멘트를 달아봤어요. 
1단계에서는 충분히 잘 해주셨기에 코멘트 반영은 2단계에서 부탁드릴게요.
추가적인 궁금증은 이 pr이 머지되어도 댓글 남겨주셔도 좋구 DM도 좋습니다.
